### PR TITLE
raft: add Status interface

### PR DESCRIPTION
**Pull request by xiang90**
Status returns the current status of raft state machine.

### CODE REVIEW

**Comment by bdarnell** [comment]
LGTM. We've also been interested in asking raft for the current node progress map (when the current node is the leader) so we can detect when a node is behind and may need to be replaced.

**Comment by xiang90** [comment]
@bdarnell Yes. Our plan is to have /local status(which is the information about self) and /peers status (which are the information about the progress of the followers).

**Comment by bdarnell** [comment]
Now that I think about it, why is Status separate from SoftState? Could we just add fields to SoftState (or replace Ready.SoftState with a Status)?

**Comment by xiang90** [comment]
@bdarnell SoftState is pushed to application. Status is pulled from raft. If there is no input, there will be no new softState. So you have to either always keep softstate in application or use something like status.

**Comment by bdarnell** [comment]
We could add a pull interface for SoftState as well. If they contain the same information it should use the same struct. The question is whether we want to push out a new Ready whenever any of the Status fields change or if some status information should be pull-only. Personally I'd prefer to use push as much as possible - we'll be managing lots of raft instances so we'd rather get changes pushed to us instead of periodically looping over all of them to collect status.

**Comment by xiang90** [comment]
@bdarnell For debugging purpose, we probably would like to pull out status at any given time even if there is no new input that can generate output. You can try to combine the structs as long as it is still clear.

FYI: we are not actively work on this issue. we will be back to it next week or so.

**Comment by xiang90** [comment]
@bdarnell refresh. Add progress into status.

Status = raft.SoftState + raft.HardState + Progress + other uncovered fields (Applied, ID).

We can try to embed `Status` into `raft struct` in the future.

It is used for debugging and serving raft stats.

File: raft/status.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+type Status struct {
+	ID uint64
+
+	pb.HardState
+	SoftState
+
+	Applied  uint64
+	Progress map[uint64]progress
```
Comment (bdarnell) [inline raft/status.go]: The progress struct (and its fields) need to be exported for this to be useful.

**Comment by bdarnell** [comment]
LGTM

File: raft/status.go (hunk 1 above)
Comment (xiang90) [inline raft/status.go]: good point.

**Pull request merged**